Use native JSON parsing when available

jakejx opened the pull request:
I took a stab at making tide use the native json parser when it is available. There is a new variable `tide-native-json-parsing` that controls which JSON parser is used. I am quite new to writing elisp so I am happy to have feedback on how to improve the code.

Closes #374

josteink commented on tide.el:
```diff
@@ -239,6 +239,14 @@ this variable to non-nil value for Javascript buffers using `setq-local' macro."
+  :type 'boolean
+  :group 'tide)
+
 (defconst tide--minimal-emacs
```
I like this. Make it configurable, at least as long as it's not properly tested.

ananthakumaran replied:
agree, I think we should keep this disabled by default for some time, and then once well tested, we can enable based on emacs version and function availability.

josteink commented on tide.el:
```diff
@@ -336,14 +344,21 @@ buffers, as they don't set `buffer-file-name' correctly."
+            (json-parse-buffer :object-type 'plist :null-object json-null :false-object json-false))
+        (let ((json-object-type 'plist))
+          (json-read-file filename)))
     (error '())))
```
Looks reasonable, but would be even better with a test-case attached, testing for both modes.

ananthakumaran replied:
should we think about adding `tide-json-encode, tide-json-decode` instead of inlining the native support check everywhere? It would be easier to debug as well if there are any differences (just encode/decode via both method and compare the values)

josteink replied:
I was considering that too, but decided that we could live with 2 repetitions of the same boilerplate, without going crazy with intermediate functions.

If however, as you say, there are even more usage (`json-read-object` inside `tide-decode-response`), abstracting it into its own function definitely makes sense.

josteink replied:
I was considering that too, but decided that we could live with 2 repetitions of the same boilerplate, without going crazy with intermediate functions.

If however, as you say, there are even more usage (`json-read-object` inside `tide-decode-response`), abstracting it into its own function definitely makes sense.

And that will also help testability.

josteink commented on tide.el:
```diff
@@ -336,14 +344,21 @@ buffers, as they don't set `buffer-file-name' correctly."
+      (if tide-native-json-parsing
+          (json-parse-string string :object-type 'plist :null-object json-null :false-object json-false)
+        (let ((json-object-type 'plist))
+          (json-read-from-string string)))
```
Same for this one: tests would be nice.

josteink commented on tide.el:
```diff
@@ -591,7 +606,9 @@ Offset is one based."
+         (encoded-command (if tide-native-json-parsing
+                              (json-serialize command :null-object json-null :false-object json-false)
+                            (json-encode command)))
          (payload (concat encoded-command "\n")))
```
It would be nice if this was also isolated in its own `defun`, and then it would be easy to add tests covering it.

josteink commented on tide.el:
```diff
@@ -1561,7 +1578,7 @@ This function is used for the basic completions sorting."
                           `(:entryNames [(:name ,name :source ,source)])
-                        `(:entryNames (,name))))
+                        `(:entryNames [,name])))
          (arguments (-concat (get-text-property 0 'file-location name)
```
This change stands out. Was it intentional? Could you elaborate on why it was required?

jakejx replied:
Yeap it was intentional. The two changes are due to `json-serialize` (native) only serialises vectors to JSON arrays (AFAIK), while`json-encoded` (elisp) seraliases both vectors and lists, so I decided to use a vector for compatibility.

josteink commented on tide.el:
```diff
@@ -2439,12 +2456,11 @@ current buffer."
    "documentHighlights"
    `(:file ,(tide-buffer-file-name) :line ,(tide-line-number-at-pos) :offset ,(tide-current-offset)
-     :filesToSearch (,(tide-buffer-file-name)))
+     :filesToSearch [,(tide-buffer-file-name)])
```
Same for this change: It stands out. Was it intentional? Why it was required?

josteink commented:
Hey there and thanks for the patch!

I've taken a look and it mostly looks good. No major objections to the implementation as it is, but I've left a few comments none the less.

If we got those resolved, I'm pretty sure we can get this merged without too many issues :+1:

jakejx commented:
Thanks for the swift review! I'll look into adding the tests 😄

ananthakumaran commented:
Most of the performance benefit is going to come from switching the `json-read-object` inside `tide-decode-response`, which I don't see in this PR. This is the only place where we decode large json responses (for autocomplete mostly) from tsserver.

jakejx commented:
> Most of the performance benefit is going to come from switching the `json-read-object` inside `tide-decode-response`, which I don't see in this PR. This is the only place where we decode large json responses (for autocomplete mostly) from tsserver.

I must have missed it, will work on using the native decoder for that.

jakejx commented:
I have updated the code to use the native JSON parser for reading objects. Currently there are 4 functions that are entry points (and thus where the check for the native flag occurs): `tide-safe-json-read-file`, `tide-safe-json-read-string`, `tide-json-read-object` and `tide-json-encode`.

I have added a single test file (just a random tsconfig) that covers the common JSON structures. The tests basically reads this file and ensures that both JSON parsers produce the same output.

One thing to note is that when reading JSON, JSON arrays are represented as lists. However, when serialising to JSON, the arrays must be vectors (as the native parser only accepts vectors). While it is possible to read JSON arrays as vectors, I think this would require other changes to other parts of the code, as `car` and `cdr` are used for list manipulation.

bimbiltu commented on tide.el:
```diff
@@ -239,6 +239,14 @@ this variable to non-nil value for Javascript buffers using `setq-local' macro."
   :group 'tide
   :safe #'booleanp)
 
+(defcustom tide-native-json-parsing (and (<= emacs-major-version 27)
```
Shouldnt this line be using `>=` instead of `<=`?

jakejx replied:
yes that was a mistake, updated.

josteink commented:
Thanks for the updates. The build seems to be failing for 2 reasons:

* The build for `emacs-git-master-snapshot` seems not to be running an old snapshot, and has had test-failures before. Not a big deal (for this PR). I'd like to look into that independently though.
* You are running your new tests on older Emacs-versions where the new JSON engine isn't present.

Could you preface your JSON-tests with `(when (>= emacs-major-version 27) ... )` ? That should fix the build and make us ready to merge, unless there are other objections.

josteink commented:
If you rebase this PR on master (or just merge latest changes from master now), the CI should be updated to include an up-to-date git master Emacs version.

This version should hopefully be able to run and verify your tests, and if that build is green, I think we can agree the tests are properly good.

jakejx commented:
It looks like the snapshot is not sufficiently up to date. This is from travis: `/home/travis/build/ananthakumaran/tide/tide.el:2806:1:Warning: the following
    functions are not known to be defined: json-parse-buffer,
    json-parse-string, json-serialize`

josteink commented:
> It looks like the native functions are not being loaded from some reason in the CI. This is from travis: `/home/travis/build/ananthakumaran/tide/tide.el:2806:1:Warning: the following functions are not known to be defined: json-parse-buffer, json-parse-string, json-serialize`

I agree. From the build-outpout:

````
$ emacs --version

GNU Emacs 27.0.50

Copyright (C) 2019 Free Software Foundation, Inc.
````

That does not seem to be an up to date build, and having that build failing does not block the merging of this PR.

josteink commented on tide.el:
```diff
@@ -239,6 +239,15 @@ this variable to non-nil value for Javascript buffers using `setq-local' macro."
+                                         (functionp 'json-serialize)
+                                         (functionp 'json-parse-buffer)
+                                         (functionp 'json-parse-string)
+                                         nil)
```
What's the point of this whole setup, if we terminate with `nil` ? Is it for making it easier to correctly switch default in the future?

jakejx replied:
yeap the main reason is to keep the native parsing disabled by default (even if the user's system supports it). Eventually we can remove the `nil` and the value will be determined by the user's system.

sandinmyjoints replied:
Maybe time to remove this?

ananthakumaran replied:
removed it in master.

josteink commented:
Trying to run this locally, based on Emacs `git master` I couldn't get any of the new tests to pass.

Looking into it, after looking quite a few places, I finally found I had a missing prerequisite: `libjansson` was not installed on my system, so obviously Emacs could not use it.

If it can be "hard" for someone even involved in `tide` to correctly implement/enable a crucial performance-related component like this, would it make sense to have this more clearly highlighted in the `README`?

Basically just a section about performance, which informs that `tide` performs better when running on Emacs-builds which has `libjansson`-use enabled?

Any opinions?

josteink commented:
You know, I've tested it manually, and things `seems` to run faster. I'll be happy to merge this code as is.

If someone wants to improve the readme, w.r.t. to the optimal setup for performance and features, that can be done in a separate PR.

Consider this merged!

jakejx commented:
> If it can be "hard" for someone even involved in `tide` to correctly implement/enable a crucial performance-related component like this, would it make sense to have this more clearly highlighted in the `README`?
> 
> Basically just a section about performance, which informs that `tide` performs better when running on Emacs-builds which has `libjansson`-use enabled?
> 
> Any opinions?

I agree with this. I will look into opening a PR to update the `README`.

josteink commented:
On a side-note: Installing `libjansson` and rebuilding Emacs, now makes `lsp-mode` fly compared to before as well. My entire Emacs feels snappier.

Thanks for helping me make my Emacs-setup even better! :smiley:
